### PR TITLE
Support async recursive calls (as long as they have indirection)

### PR DESCRIPTION
**Pull request by compiler-errors**
Before #101692, we stored coroutine witness types directly inside of the coroutine. That means that a coroutine could not contain itself (as a witness field) without creating a cycle in the type representation of the coroutine, which we detected with the `OpaqueTypeExpander`, which is used to detect cycles when expanding opaque types after that are inferred to contain themselves.

After `-Zdrop-tracking-mir` was stabilized, we no longer store these generator witness fields directly, but instead behind a def-id based query. That means there is no technical obstacle in the compiler preventing coroutines from containing themselves per se, other than the fact that for a coroutine to have a non-infinite layout, it must contain itself wrapped in a layer of allocation indirection (like a `Box`).

This means that it should be valid for this code to work:

```rust
async fn async_fibonacci(i: u32) -> u32 {
    if i == 0 || i == 1 {
        i
    } else {
        Box::pin(async_fibonacci(i - 1)).await 
          + Box::pin(async_fibonacci(i - 2)).await
    }
}
```

Whereas previously, you'd need to coerce the future to `Pin<Box<dyn Future<Output = ...>>` before `await`ing it, to prevent the async's desugared coroutine from containing itself across as await point.

This PR does two things:
1. Only report an error if an opaque expansion cycle is detected *not* through coroutine witness fields.
    * Instead, if we find an opaque cycle through coroutine witness fields, we compute the layout of the coroutine. If that results in a cycle error, we report it as a recursive async fn.
4. Reworks the way we report layout errors having to do with coroutines, to make up for the diagnostic regressions introduced by (1.). We actually do even better now, pointing out the call sites of the recursion!

### CODE REVIEW

**Comment by compiler-errors** [comment]
:(

Perf regression probably has to do with the fact that we compute the layout of the opaque type eagerly. Modified strategy slightly.

@bors try @rust-timer queue

**Comment by compiler-errors** [comment]
I think this is ready for review before fcp. cc @rust-lang/wg-async

File: compiler/rustc_error_codes/src/error_codes/E0733.md [hunk 1]
```diff
@@ -13,7 +13,7 @@ async fn foo(n: usize) {
 To perform async recursion, the `async fn` needs to be desugared such that the
```
Comment (lcnr) [inline compiler/rustc_error_codes/src/error_codes/E0733.md:13]: It feels preferable to change the suggestion to suggest `Box::pin(recur()).await` as the first option (noting that it requires a newer version), this is a fairly significant improvement

File: compiler/rustc_hir/src/hir.rs [hunk 2]
```diff
@@ -1529,6 +1529,12 @@ pub enum CoroutineKind {
     Coroutine,
 }
 
+impl CoroutineKind {
+    pub fn is_fn_like(self) -> bool {
+        matches!(self, CoroutineKind::Async(CoroutineSource::Fn) | CoroutineKind::Gen(CoroutineSource::Fn))
```
Comment (lcnr) [inline compiler/rustc_hir/src/hir.rs]: nit: can you use an exhaustive match here?

File: compiler/rustc_middle/src/ty/util.rs [hunk 3]
```diff
@@ -850,18 +850,7 @@ impl<'tcx> OpaqueTypeExpander<'tcx> {
...
-                    for bty in self.tcx.coroutine_hidden_types(def_id) {
-                        let hidden_ty = bty.instantiate(self.tcx, args);
-                        self.fold_ty(hidden_ty);
-                    }
-                    let expanded_ty = Ty::new_coroutine_witness(self.tcx, def_id, args);
-                    self.expanded_cache.insert((def_id, args), expanded_ty);
-                    expanded_ty
-                }
-            };
+            let expanded_ty = Ty::new_coroutine_witness(self.tcx, def_id, args);
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/util.rs]: I would expect us to completely remove this function 🤔 `expanded_ty` is the same type as the input and I don't think there's any use to return `None` if there's any recursion 🤔

given that we don't expand coroutine witnesses anymore, we already handly cycles involving coroutines via the ordinary occurs check. Any cyclic coroutine would have to result in a type of infinite size.

File: compiler/rustc_trait_selection/src/traits/query/normalize.rs [hunk 4]
```diff
@@ -241,16 +241,13 @@ impl<'cx, 'tcx> FallibleTypeFolder<TyCtxt<'tcx>> for QueryNormalizer<'cx, 'tcx>
...
                         self.anon_depth += 1;
                         if concrete_ty == ty {
-                            bug!(
-                                "infinite recursion generic_ty: {:#?}, args: {:#?}, \
-                                 concrete_ty: {:#?}, ty: {:#?}",
-                                generic_ty,
-                                args,
-                                concrete_ty,
-                                ty
+                            concrete_ty = Ty::new_error_with_message(
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/query/normalize.rs:247]: why did this change from a `bug` to a `delay_span_bug`, how did the other changes in this commit cause this to be reachable?

File: compiler/rustc_middle/src/values.rs [hunk 5]
```diff
@@ -154,13 +154,37 @@ impl<'tcx, T> Value<TyCtxt<'tcx>> for Result<T, &'_ ty::layout::LayoutError<'_>>
...
-                .emit()
+            let mut diag = struct_span_err!(
+                tcx.sess,
+                span,
+                E0733,
+                "recursion in an `async fn` requires boxing"
+            );
+            diag.note("a recursive `async fn` call must introduce indirection such as `Box::pin` to avoid an infinitely sized future");
+            diag.note(
+                "consider using the `async_recursion` crate: https://crates.io/crates/async_recursion",
```
Comment (lcnr) [inline compiler/rustc_middle/src/values.rs]: do we still want to suggest the `async_recursion` crate. manually changing the recursive call to `Box::pin(recur()).await)` feels preferable to adding a new dependency. In my experience suggesting multiple solutions in diagnostics causes them to be less helpful as they end up somewhat overwhelming the user, causing them to ignore the diagnostic entirely 🤔

File: compiler/rustc_hir_analysis/src/check/check.rs [hunk 6]
```diff
@@ -235,16 +235,33 @@ pub(super) fn check_opaque_for_cycles<'tcx>(
...
     let args = GenericArgs::identity_for_item(tcx, def_id);
-    if tcx.try_expand_impl_trait_type(def_id.to_def_id(), args).is_err() {
-        let reported = opaque_type_cycle_error(tcx, def_id, span);
-        Err(reported)
-    } else if let Err(&LayoutError::Cycle(guar)) =
-        tcx.layout_of(tcx.param_env(def_id).and(Ty::new_opaque(tcx, def_id.to_def_id(), args)))
+
+    // First, try to look at any opaque expansion cycles, considering coroutine fields
+    // (even though these aren't necessarily true errors).
+    if tcx
```
Comment (lcnr) [inline compiler/rustc_hir_analysis/src/check/check.rs:241]: could we alternatively completely skip checking coroutines here? 🤔 it feels a bit off for me to mix coroutine and opaque type cycle checking as they are pretty much fundamentally different checks

e.g. i could imagine a separate global query check which eagerly computes the layout of coroutines

**Review by lcnr — COMMENTED** [review]
incredibly happy about this change. It's a major improvement 🎉 

a few comments and nits, r=me after t-types fcp afterwards

File: compiler/rustc_middle/src/values.rs (hunk 5 above)
Comment (lcnr) [reply compiler/rustc_middle/src/values.rs]: @rust-lang/wg-async relevant for you: if simply using `Box::pin` works now, should we still suggest to use the `async_recursion` macro?

File: compiler/rustc_middle/src/ty/util.rs (hunk 3 above)
Comment (compiler-errors) [reply compiler/rustc_middle/src/ty/util.rs]: Yes, but it's cheaper to only do this check once on the "normal" path, and only compute the layout of the coroutine if it's detected to be cyclic.

See the perf regression up-thread and the PR that fixed it.

File: compiler/rustc_trait_selection/src/traits/query/normalize.rs (hunk 4 above)
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/query/normalize.rs:247]: It's because we're normalizing infinitely sized opaques in `layout_of`.

**Comment by lcnr** [comment]
tried an alternative approach in #118136 and think we should go with this approach. However, can you integrate the improvement to the `impl<'tcx, T> Value<TyCtxt<'tcx>> for Result<T, &'_ ty::layout::LayoutError<'_>>` from that PR?

I generally think that `from_cycle_error` isn't ideal and we should do the "search for the optimal cycle head" for all cycle errors :thinking: but that's for a different PR

Personally in favor of removing the mention of the `async_recursion`, would start the FCP after that?

**Comment by yoshuawuyts** [comment]
I like this change in principle since it would make async recursion easier to perform. I hope this will lessen the need for crates like [async-recursion](https://docs.rs/async-recursion/latest/async_recursion/) to be used. For completion, here is the same code written three times, side-by-side for comparison.

## Using Rust 1.74 stable

This is the way we would write the code on stable today, without any crates:

```rust
use core::future::Future;
use core::pin::Pin;

fn async_fibonacci(i: u32) -> Pin<Box<dyn Future<Output = u32>>> {
    Box::pin(async move {
        if i == 0 || i == 1 {
            i
        } else {
            async_fibonacci(i - 1).await 
              + async_fibonacci(i - 2).await
        }
    })
}
```

## Using `async-recursion`

Using the `async-recursion` crate. This has the downside that each invocation now returns a `Pin<Box<_>>`, which strictly means more allocations than required since the first call to a recursive function does not need to be boxed.


```rust
#[async_recursion(?Send)]
async fn async_fibonacci(i: u32) -> u32 {
    if i == 0 || i == 1 {
        i
    } else {
        async_fibonacci(i - 1).await 
          + async_fibonacci(i - 2).await
    }
}
```

## Using this patch

And here is the same code again using this patch. This is more efficient than either variant. And while not quite as ergonomic as `async-recursion`, it seems like a meaningful improvement compared to the status quo without libraries today. It's also worth noting that accepting this change doesn't seem like it would close the doors on any further improvements to async recursion in the future. Meaning I see very little reason not to accept this from a language perspective.

```rust
async fn async_fibonacci(i: u32) -> u32 {
    if i == 0 || i == 1 {
        i
    } else {
        Box::pin(async_fibonacci(i - 1)).await 
          + Box::pin(async_fibonacci(i - 2)).await
    }
}
```

**Comment by compiler-errors** [comment]
I'll kick off the fcp now. See PR description for detail.

For T-lang: Do you agree that these semantics (helpfully demonstrated by yosh in https://github.com/rust-lang/rust/pull/117703#issuecomment-1826352494) are desirable to commit to?

For T-types: Do we feel comfortable with relying on the design of `GeneratorWitness` (formerly `GeneratorWitnessMIR`) to allow us to represent coroutines that contain themselves modulo indirection?

@rfcbot fcp merge

**Comment by RalfJung** [comment]
> Using the async-recursion crate. This has the downside that each invocation now returns a Pin\<Box\<_>>, which strictly means more allocations than required since the first call to a recursive function does not need to be boxed.

This downside is mentioned for the async-recursion  variant but not the stable variant. But I assume it applies to both?

**Comment by yoshuawuyts** [comment]
@RalfJung yeah that's right. In hindsight I should have mentioned that.

**Comment by lcnr** [comment]
checked the box for @oli-obk as they are on leave:

> Just check my box. I'll review my notifications at some point

**Comment by tmandry** [comment]
I'm really happy for this change!

@rfcbot reviewed

**Comment by aliemjay** [comment]
@rfcbot reviewed

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

We discussed this in the T-lang meeting today.  There was no specific consensus other than that people would have a look at this.  Enough people now have that this is in FCP.  So we'll remove the nomination.

File: compiler/rustc_middle/src/values.rs [hunk 7]
```diff
@@ -102,19 +111,107 @@ impl<'tcx> Value<TyCtxt<'tcx>> for Representability {
...
 impl<'tcx, T> Value<TyCtxt<'tcx>> for Result<T, &'_ ty::layout::LayoutError<'_>> {
-    fn from_cycle_error(_tcx: TyCtxt<'tcx>, _cycle: &[QueryInfo], guar: ErrorGuaranteed) -> Self {
+    fn from_cycle_error(
+        tcx: TyCtxt<'tcx>,
+        cycle_error: &CycleError,
+        _guar: ErrorGuaranteed,
+    ) -> Self {
+        let mut cycle: VecDeque<_> = cycle_error.cycle.iter().collect();
+
+        let guar = 'search: {
```
Comment (lcnr) [inline compiler/rustc_middle/src/values.rs:141]: can you move this search into a subfunction, rn it's somewhat difficult to read imo

File: compiler/rustc_error_codes/src/error_codes/E0733.md (hunk 1 above)
Comment (lcnr) [reply compiler/rustc_error_codes/src/error_codes/E0733.md:13]: while I would still like to change it to use `Box::pin` as the first suggestion, this is not blocking :shrug:

**Review by lcnr — COMMENTED** [review]
r=me after nit

File: compiler/rustc_error_codes/src/error_codes/E0733.md [hunk 8]
```diff
@@ -10,21 +10,22 @@ async fn foo(n: usize) {
...
 
-To perform async recursion, the `async fn` needs to be desugared such that the
-`Future` is explicit in the return type:
+The recursive invocation can be boxed:
 
-```edition2018,compile_fail,E0733
-use std::future::Future;
-fn foo_desugared(n: usize) -> impl Future<Output = ()> {
+```edition2018
+async fn foo(n: usize) {
```
Comment (lcnr) [inline compiler/rustc_error_codes/src/error_codes/E0733.md:16]:
```suggestion
async fn foo_recursive(n: usize) {
```

Comment (compiler-errors) [reply compiler/rustc_error_codes/src/error_codes/E0733.md:16]: I'd rather just call everything `foo`

Comment (lcnr) [reply compiler/rustc_error_codes/src/error_codes/E0733.md:16]: also good

**Comment by compiler-errors** [comment]
rebased

@bors r=lcnr

**Comment by pnkfelix** [comment]
* this is weird, it looks like an inverse blip occurred on the preceding PR (#119760), where tt-muncher check incr-unchanged had a single point with -0.2% instruction-count, and then it preceding to "return to normal" on the succeeding PRs.
* (Its harder for me to explain away "inverse blips" ...)
* but at the same time, this does not seem like a significant regression by our usual metrics.
* marking as triaged.

@rustbot label: +perf-regression-triaged

**Pull request merged**